### DISCUSSION TITLE
how do you manually position tooltips?

### DISCUSSION THREAD

**Discussion by twolf2919** [jfree/jfreechart#347] [Q&A] [answered]
I first did a google search and came across this stack overflow thread:
[https://stackoverflow.com/questions/23698417/java-jfreechart-customize-tooltip-screen-position](url) in which the accepted answer seemed to be that one needs to extend ChartPanel and override the getToolTipLocation(MouseEvent) method.  I did that, but that method never gets called.  Perhaps I'm missing something?  Below are the relevant code bits:
`
		XYDataset dataset = createDataset(s1, s2, s3);
		JFreeChart chart = ChartFactory.createXYLineChart(null, X_AXIS_LABEL, logAxis.getLabel(), dataset);
		customizeChart(chart);
		JPanel panel = new ChartPanel(chart) {
			public Point getToolTipLocation(MouseEvent event) {
				System.err.println("Got here");
				Point p = super.getToolTipLocation(event);
				return new Point(p.x + 150, p.y);
			}			
		};
 	private void customizeChart(JFreeChart chart) { 
		XYPlot plot = chart.getXYPlot();
		XYLineAndShapeRenderer renderer = new XYLineAndShapeRenderer();		
		StandardXYToolTipGenerator tooltip = new StandardXYToolTipGenerator() {
			public String generateToolTip(XYDataset dataset, int series, int item) {
				Number y = dataset.getY(series, item);
				NumberFormat formatter = NumberFormat.getCurrencyInstance();
				String moneyString = formatter.format(y.doubleValue());
				return moneyString;
			}
		};
	
		// sets paint color for each series
		renderer.setSeriesPaint(0, Color.RED);
		renderer.setSeriesToolTipGenerator(0,tooltip);

`
Any suggestions/pointers would be much appreciated.

**Comment by trashgod** [comment, marked as answer]
I am unable to reproduce this. I'm not sure where your fragment goes awry, but you appear to be creating a new renderer and tooltip generator that is never used. Note that the default implementation of `getToolTipLocation()` returns `null`.

**Reply by twolf2919** [reply]
Hi - thanks for your reply.    I forgot to paste one line:
     plot.setRenderer(renderer);
Sorry about that.  It's really a mystery to me that the getToolTipLocation() is never called - I put a printf in both it and in the generator's generateTooltip() and the generator's method does get called - and my tooltips are getting displayed formatted correctly, but the ChartPanel's getToolTipLocation() never does.  I feel like I'm missing something fundamental - like is there something that needs to be set to enable this query by Java?  getToolTipLocaion() is a JComponent method.  I guess next I will give the override a try on one of my JComponents to see if it gets called there automatically.

**Reply by twolf2919** [reply]
I just verified that when I override the getToolTipLocation() method on any of my Swing-based components - e.g. Button - that method does get called.   Next, I found another place in my app where I create a chart - this one is a Pie chart created via:
		JFreeChart pieChart = ChartFactory.createPieChart(title, dataset, false, true, false);
and when I override the method on its ChartPanel, te overridden method does get called.  So maybe it's something  unique to the XYLineChart or the fact that here I'm also using a renderer (in the pie chart I don't)?

**Reply by twolf2919** [reply]
I'm terribly sorry to waste everyone's time who read this - it turns out the bit of code where I was adding the tooltip position override wasn't actually the bit of code that executed - someone had duplicated my routine, so my routine wasn't even getting called :-(  Again, many apologies.  Everything works fine now - I was able to override and move the tooltip position out of the way of a larger cursor.

**Comment by trashgod** [comment]
I had no problem with this [example](https://stackoverflow.com/a/71801198/230513) and the following implementation. Note that it is called only _after_ the tooltip is displayed.

```
    @Override
    public Point getToolTipLocation(MouseEvent e) {
        System.out.println(e);
        return super.getToolTipLocation(e);
    }
```

I've always relied on the default positioning, but this Swing [topic](https://stackoverflow.com/search?tab=votes&q=%5bswing%5d%20getToolTipLocation&searchOn=3) may offer some insight. As an aside, the example discussed [here](https://stackoverflow.com/q/64217855/230513) and seen [here](https://gist.github.com/trashgod/a3f7581417db19bf5b6dc52d3647a2f9) uses an `XYTextAnnotation` to mimic a static tooltip until mouse movement triggers the actual display.

**Reply by twolf2919** [reply]
I guess I don't follow how the example you point to is related to tooltips.  The override you gave is essentially the same as mine - but for me the println never happens.

I *want* to rely on the default positioning, but on my Mac I use a slightly larger mouse pointer - not something I do within the program but something I set in the System preferences to make it easier to discover where my cursor is on a 3 monitor setup :-) - and the bottom-right of the mouse pointer *always* obscures the tooltip!   It seems a shortcoming in Java as I see the same issue in my other Tooltip toting components.

Since you can't reproduce it, I will now try a very basic Java program with the above chart and see if it occurs there - perhaps it's got to do with the depth of my component tree or other app-specific doings.  Thanks for the help so far.

**Reply by trashgod** [reply]
It looks like you got it sorted above. For the particular case of locating the cursor on macOS, also consider the accessibility feature mentioned [here](https://apple.stackexchange.com/questions/189006/mouse-pointer-locator).
